### ISSUE TITLE
Utility to change the icon

### ISSUE THREAD

**Issue by TrueCat17** [erocarrera/pefile#375]
Many cross-platform engines and applications need to set the icon for the *.exe file regardless of the platform.
Judging by the fact that almost everywhere Windows or Wine is required to change an icon, there is simply no solution to this problem.

It seems to be a common and not so difficult task, but all over the Internet I found exactly 1 way to do this (and even then - the result is problematic).
There is a lot of information about how to extract information (including the icon), how to change the file with third-party utilities or with their partial use, but nothing about how you can simply change the icon without win-api.

The way I found (from the renpy/launcher project) uses your library, which is why I'm referring here.
change_icon.py - edited (without internal dependencies) file from renpy.
`python change_icon.py old.exe icon.ico new.exe`

In my case, the resulting file has the desired icon, runs in Wine, but (suddenly) doesn't run in Windows.

Uncommenting the `newpe.print_info()` line at the very beginning produces the following warning:
`Invalid relocation information. VirtualAddress outside of Image: 0x688ef2f1`

Regarding the same output, `pe` (original exe) immediately shows that the part with `Base relocations` (at the bottom of the output) is simply missing.
I tried to fix it by adding the last couple of commented lines.
After that, the part with relocations reappeared, but now the file does not run anywhere at all.

This is not the first time I spend several days trying to solve this problem, but nothing comes of it, because I am very poorly versed in all these tables, sections, headers, sizes, addresses, offsets, etc.

Perhaps there is a person here who can once and for all solve this problem for the entire Internet?
Of course, it would be nice to make a separate python module for this, rather than just posting an answer here, but this is no longer relevant to the current topic.
Thanks.

[change_icon.zip](https://github.com/erocarrera/pefile/files/11537247/change_icon.zip)

**Comment by j-t-1** [comment]
Probably the easiest way is to replace the icons with new icons that are equivalent size.

Since this issue was posted, a link to icoextract was added to pefile, and this has some functionality that can help.

Using `icolist` the file offset of the first icon is found, and it is extracted using and `icoextract`.

Below was going to be a very basic example to show the idea, although it does not work.

```
from PIL import Image

# icolist Fondue.exe
file_offset = 0xb808

# icoextract Fondue.exe fondue.ico
icon = "fondue.ico"

with Image.open(icon) as img:
    rotated = img.rotate(180, expand=False)
    rotated.save("rotated.ico", format="ICO")

with open("rotated.ico", 'rb') as f:
    new_icon = f.read() 

with open("Fondue.exe", "r+b") as f:
    f.seek(file_offset)
    f.write(new_icon)
    f.flush()
```

I think that all the different resolution of icons that can be used for the file icon need to be changed but am unsure how to distinguish these from icons that would be used for other purposes.

@jlu5 would you be able to help us?
Does the icon extracted by icoextract correspond to the first icon in icolist?
How to identify which icons need to be replaced using icolist?

**Comment by TrueCat17** [comment]
It's just a workaround.
It can be used, of course, but it's not a bug fix.

One problem is that it's unclear what to do if the icon isn't there, or if the new one is larger.

The second problem is that a 256x256 icon will weigh 0.5 MB uncompressed, and even more if you include smaller versions.
And you'll have to carry it around, and File Explorer will have to read it all every time to display the icon.
On older hard drives, I think this could create additional lag.

**Comment by j-t-1** [comment]
Yes is a work around; if the icon is a different size to the existing, resizing to the existing size would be needed.

If the icon does not exist, Resource Hacker seems to be able to add one.

**Comment by jlu5** [comment]
I don't think patching the binary directly is the right approach. You probably want a resource aware patcher like Resource Hacker or https://github.com/avast/pe_tools

The latter has [an example](https://github.com/avast/pe_tools#examples) on how to replace an existing resource entry. Judging by the description, you probably just need the target icon's resource ID (which icolist can find) and a local copy of the new .ico file. Outside Windows you can use MinGW (x86_64-w64-mingw32-windres) or Wine (wrc) instead of rc.exe to turn the .rc resource script into a .res file.

**Comment by TrueCat17** [comment]
I came across pe_tools, but discarded it because I thought it would be useless on Linux without Wine.
If mingw32-windres can replace rc.exe, then it would all make sense.
And if this utility can work separately from the rest of mingw-binutils and can be ported to Python, then that would be fantastic.

**Comment by jlu5** [comment]
> And if this utility can work separately from the rest of mingw-binutils and can be ported to Python, then that would be fantastic.

Why reinvent the wheel? It's a CLI program which is easy to instrument in a subprocess.

**Comment by TrueCat17** [comment]
Basically, I'm making a standalone utility that needs to be cross-platform and work simply by unpacking it from a zip archive.
Building for all the necessary platforms or having to install anything else is a big disadvantage for my situation.

**Comment by j-t-1** [comment]
Adding a link in this issue may help find potential contributors.
